File: FbxSharpTests/FbxDocumentInfoTest.cs
using System;
using NUnit.Framework;
using FbxSharp;

namespace FbxSharpTests
{
    [TestFixture]
    public class FbxDocumentInfoTest : TestBase
    {
        [Test]
        public void FbxDocumentInfo_Create_HasDefaults()
        {
            // given:
            var dt0 = new FbxDateTime();
            FbxProperty prop;

            // when:
            var docinfo = FbxDocumentInfo.Create("");

            // then:
            Assert.AreEqual(15, CountProperties(docinfo));

            prop = docinfo.FindProperty("DocumentUrl");
            Assert.True(prop.IsValid());
            Assert.AreEqual("DocumentUrl", prop.GetName());
            Assert.AreEqual("DocumentUrl", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxUrlDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.LastSavedUrl);

            prop = docinfo.FindProperty("SrcDocumentUrl");
            Assert.True(prop.IsValid());
            Assert.AreEqual("SrcDocumentUrl", prop.GetName());
            Assert.AreEqual("SrcDocumentUrl", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxUrlDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.Url);

            prop = docinfo.FindProperty("Original");
            Assert.True(prop.IsValid());
            Assert.AreEqual("Original", prop.GetName());
            Assert.AreEqual("Original", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxCompoundDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.Original);

            prop = docinfo.FindPropertyHierarchical("Original|ApplicationVendor");
            Assert.True(prop.IsValid());
            Assert.AreEqual("ApplicationVendor", prop.GetName());
            Assert.AreEqual("Original|ApplicationVendor", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxStringDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.Original_ApplicationVendor);

            prop = docinfo.FindPropertyHierarchical("Original|ApplicationName");
            Assert.True(prop.IsValid());
            Assert.AreEqual("ApplicationName", prop.GetName());
            Assert.AreEqual("Original|ApplicationName", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxStringDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.Original_ApplicationName);

            prop = docinfo.FindPropertyHierarchical("Original|ApplicationVersion");
            Assert.True(prop.IsValid());
            Assert.AreEqual("ApplicationVersion", prop.GetName());
            Assert.AreEqual("Original|ApplicationVersion", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxStringDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.Original_ApplicationVersion);

            prop = docinfo.FindPropertyHierarchical("Original|DateTime_GMT");
            Assert.True(prop.IsValid());
            Assert.AreEqual("DateTime_GMT", prop.GetName());
            Assert.AreEqual("Original|DateTime_GMT", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxDateTimeDT, prop.GetPropertyDataType());
            Assert.AreEqual(dt0, prop.Get<FbxDateTime>());
            Assert.True(prop == docinfo.Original_DateTime_GMT);

            prop = docinfo.FindPropertyHierarchical("Original|FileName");
            Assert.True(prop.IsValid());
            Assert.AreEqual("FileName", prop.GetName());
            Assert.AreEqual("Original|FileName", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxStringDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.Original_FileName);

            prop = docinfo.FindProperty("LastSaved");
            Assert.True(prop.IsValid());
            Assert.AreEqual("LastSaved", prop.GetName());
            Assert.AreEqual("LastSaved", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxCompoundDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.LastSaved);

            prop = docinfo.FindPropertyHierarchical("LastSaved|ApplicationVendor");
            Assert.True(prop.IsValid());
            Assert.AreEqual("ApplicationVendor", prop.GetName());
            Assert.AreEqual("LastSaved|ApplicationVendor", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxStringDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.LastSaved_ApplicationVendor);

            prop = docinfo.FindPropertyHierarchical("LastSaved|ApplicationName");
            Assert.True(prop.IsValid());
            Assert.AreEqual("ApplicationName", prop.GetName());
            Assert.AreEqual("LastSaved|ApplicationName", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxStringDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.LastSaved_ApplicationName);

            prop = docinfo.FindPropertyHierarchical("LastSaved|ApplicationVersion");
            Assert.True(prop.IsValid());
            Assert.AreEqual("ApplicationVersion", prop.GetName());
            Assert.AreEqual("LastSaved|ApplicationVersion", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxStringDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.LastSaved_ApplicationVersion);

            prop = docinfo.FindPropertyHierarchical("LastSaved|DateTime_GMT");
            Assert.True(prop.IsValid());
            Assert.AreEqual("DateTime_GMT", prop.GetName());
            Assert.AreEqual("LastSaved|DateTime_GMT", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxDateTimeDT, prop.GetPropertyDataType());
            Assert.AreEqual(dt0, prop.Get<FbxDateTime>());
            Assert.True(prop == docinfo.LastSaved_DateTime_GMT);

            prop = docinfo.FindProperty("DocumentEmbeddedUrl");
            Assert.True(prop.IsValid());
            Assert.AreEqual("DocumentEmbeddedUrl", prop.GetName());
            Assert.AreEqual("DocumentEmbeddedUrl", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxUrlDT, prop.GetPropertyDataType());
            Assert.AreEqual("", prop.Get<string>());
            Assert.True(prop == docinfo.EmbeddedUrl);

            prop = docinfo.FindProperty("SceneThumbnail");
            Assert.True(prop.IsValid());
            Assert.AreEqual("SceneThumbnail", prop.GetName());
            Assert.AreEqual("SceneThumbnail", prop.GetHierarchicalName());
            Assert.AreEqual(FbxDataTypes.FbxReferenceObjectDT, prop.GetPropertyDataType());
            Assert.AreEqual(null, prop.Get<FbxObject>());
        }
    }
}

File: FbxCppTests/FbxDocumentInfoTest.cpp

#include "Tests.h"

using namespace std;

void FbxDocumentInfo_Create_HasDefaults()
{
    // given:
    FbxManager* manager = FbxManager::Create();
    FbxDateTime dt0 = FbxDateTime();
    FbxProperty prop;

    // when:
    FbxDocumentInfo* docinfo = FbxDocumentInfo::Create(manager, "");

    // then:
    AssertEqual(15, CountProperties(docinfo));

    prop = docinfo->FindProperty("DocumentUrl");
    AssertTrue(prop.IsValid());
    AssertEqual("DocumentUrl", prop.GetName());
    AssertEqual("DocumentUrl", prop.GetHierarchicalName());
    AssertEqual(FbxUrlDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->LastSavedUrl);

    prop = docinfo->FindProperty("SrcDocumentUrl");
    AssertTrue(prop.IsValid());
    AssertEqual("SrcDocumentUrl", prop.GetName());
    AssertEqual("SrcDocumentUrl", prop.GetHierarchicalName());
    AssertEqual(FbxUrlDT, prop.GetPropertyDataType());
    AssertEqual(FbxString(""), prop.Get<FbxString>());
    AssertTrue(prop == docinfo->Url);

    prop = docinfo->FindProperty("Original");
    AssertTrue(prop.IsValid());
    AssertEqual("Original", prop.GetName());
    AssertEqual("Original", prop.GetHierarchicalName());
    AssertEqual(FbxCompoundDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->Original);

    prop = docinfo->FindPropertyHierarchical("Original|ApplicationVendor");
    AssertTrue(prop.IsValid());
    AssertEqual("ApplicationVendor", prop.GetName());
    AssertEqual("Original|ApplicationVendor", prop.GetHierarchicalName());
    AssertEqual(FbxStringDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->Original_ApplicationVendor);

    prop = docinfo->FindPropertyHierarchical("Original|ApplicationName");
    AssertTrue(prop.IsValid());
    AssertEqual("ApplicationName", prop.GetName());
    AssertEqual("Original|ApplicationName", prop.GetHierarchicalName());
    AssertEqual(FbxStringDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->Original_ApplicationName);

    prop = docinfo->FindPropertyHierarchical("Original|ApplicationVersion");
    AssertTrue(prop.IsValid());
    AssertEqual("ApplicationVersion", prop.GetName());
    AssertEqual("Original|ApplicationVersion", prop.GetHierarchicalName());
    AssertEqual(FbxStringDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->Original_ApplicationVersion);

    prop = docinfo->FindPropertyHierarchical("Original|DateTime_GMT");
    AssertTrue(prop.IsValid());
    AssertEqual("DateTime_GMT", prop.GetName());
    AssertEqual("Original|DateTime_GMT", prop.GetHierarchicalName());
    AssertEqual(FbxDateTimeDT, prop.GetPropertyDataType());
    AssertEqual(dt0, prop.Get<FbxDateTime>());
    AssertTrue(prop == docinfo->Original_DateTime_GMT);

    prop = docinfo->FindPropertyHierarchical("Original|FileName");
    AssertTrue(prop.IsValid());
    AssertEqual("FileName", prop.GetName());
    AssertEqual("Original|FileName", prop.GetHierarchicalName());
    AssertEqual(FbxStringDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->Original_FileName);

    prop = docinfo->FindProperty("LastSaved");
    AssertTrue(prop.IsValid());
    AssertEqual("LastSaved", prop.GetName());
    AssertEqual("LastSaved", prop.GetHierarchicalName());
    AssertEqual(FbxCompoundDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->LastSaved);

    prop = docinfo->FindPropertyHierarchical("LastSaved|ApplicationVendor");
    AssertTrue(prop.IsValid());
    AssertEqual("ApplicationVendor", prop.GetName());
    AssertEqual("LastSaved|ApplicationVendor", prop.GetHierarchicalName());
    AssertEqual(FbxStringDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->LastSaved_ApplicationVendor);

    prop = docinfo->FindPropertyHierarchical("LastSaved|ApplicationName");
    AssertTrue(prop.IsValid());
    AssertEqual("ApplicationName", prop.GetName());
    AssertEqual("LastSaved|ApplicationName", prop.GetHierarchicalName());
    AssertEqual(FbxStringDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->LastSaved_ApplicationName);

    prop = docinfo->FindPropertyHierarchical("LastSaved|ApplicationVersion");
    AssertTrue(prop.IsValid());
    AssertEqual("ApplicationVersion", prop.GetName());
    AssertEqual("LastSaved|ApplicationVersion", prop.GetHierarchicalName());
    AssertEqual(FbxStringDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->LastSaved_ApplicationVersion);

    prop = docinfo->FindPropertyHierarchical("LastSaved|DateTime_GMT");
    AssertTrue(prop.IsValid());
    AssertEqual("DateTime_GMT", prop.GetName());
    AssertEqual("LastSaved|DateTime_GMT", prop.GetHierarchicalName());
    AssertEqual(FbxDateTimeDT, prop.GetPropertyDataType());
    AssertEqual(dt0, prop.Get<FbxDateTime>());
    AssertTrue(prop == docinfo->LastSaved_DateTime_GMT);

    prop = docinfo->FindProperty("DocumentEmbeddedUrl");
    AssertTrue(prop.IsValid());
    AssertEqual("DocumentEmbeddedUrl", prop.GetName());
    AssertEqual("DocumentEmbeddedUrl", prop.GetHierarchicalName());
    AssertEqual(FbxUrlDT, prop.GetPropertyDataType());
    AssertEqual("", prop.Get<FbxString>());
    AssertTrue(prop == docinfo->EmbeddedUrl);

    prop = docinfo->FindProperty("SceneThumbnail");
    AssertTrue(prop.IsValid());
    AssertEqual("SceneThumbnail", prop.GetName());
    AssertEqual("SceneThumbnail", prop.GetHierarchicalName());
    AssertEqual(FbxReferenceObjectDT, prop.GetPropertyDataType());
    AssertEqual(NULL, prop.Get<FbxObject*>());
}

void FbxDocumentInfoTest::RegisterTestCases()
{
    AddTestCase(FbxDocumentInfo_Create_HasDefaults);
}

